### PR TITLE
Add signal handlers

### LINKED ISSUE #2339: Signal Handlers
Something equivalent of the followings:

node.js
```js
process.on('SIGTERM', () => {/* do something */}));
```

python
```py
def handler(signum, frame):
  # do something

signal.signal(signal.SIGTERM, handler)
```

ruby
```rb
Signal.trap("TERM") {
  # do something
  exit
}
```

golang
```go
sigc := make(chan os.Signal, 1)
signal.Notify(sigc, syscall.SIGTERM)
go func() {
    s := <-sigc
    // do somethig
}()
```

rust (using tokio signal)
```rust
let ctrl_c = tokio_signal::ctrl_c().flatten_stream();

let prog = ctrl_c.for_each(|()| {
    // do something
    Ok(())
});

tokio::run(prog.map_err(|err| panic!("{}", err)));
```

### PR DESCRIPTION
**Pull request by kt3k**
This PR adds signal handler APIs. `Deno.signal(signo)` function returns an async iterator of the given signal number. The returned object also works as a promise which is resolved when the first signal is received.

Streaming API
```ts
for await (const _ for Deno.signal(Deno.Signal.SIGTERM)) {
  console.log("SIGTERM received!");
}
```

Promise API
```ts
await Deno.signal(Deno.Signal.SIGTERM);
console.log("SIGTERM!");
```

The stream can be disposed by calling `.dispose()` method.
```ts
const sig = Deno.signal(Deno.Signal.SIGTERM);
setTimeout(() => { sig.dispose(); }, 5000);

for await (const _ of sig) {
  console.log("SIGTERM received!");
}
```

The above for-await-of loop exits after 5000ms.

Note: Signal handlers doesn't block the program exiting. The following program exits immediately.

```
await Deno.signal(Deno.Signal.SIGTERM);
```
Signal handlers must be used with some other main things, such as servers, etc, and those main routines should block the program from exiting. See the discussions in #3721 #3715 for details.

---

This PR also adds the following shorthand functions:
```ts
Deno.signals.alarm(); // for SIGALRM
Deno.signals.child(); // for SIGCHLD
Deno.signals.hungup(); // for SIGHUP
Deno.signals.info(); // for SIGINFO
Deno.signals.interrupt(); // for SIGINT
Deno.signals.io(); // for SIGIO
Deno.signals.pipe(); // for SIGPIPE
Deno.signals.quit(); // for SIGQUIT
Deno.signals.terminate(); // for SIGTERM
Deno.signals.userDefined1(); // for SIGUSR1
Deno.signals.userDefined2(); // for SIGURS2
Deno.signals.windowChange(); // for SIGWINCH
```
These shorthand names are inspired by methods of [`tokio::signal::unix::SignalKind`](https://tokio-rs.github.io/tokio/doc/tokio/signal/unix/struct.SignalKind.html).

---

#### Internals

This PR adds 3 ops:
- op_bind_signal (start watching signals) - Sync
- op_poll_singal (poll single signal) - [AsyncUnref](https://github.com/denoland/deno/pull/3721)
- op_unbind_signal (stop watching signals) - Sync

And 1 resource:
- SignalStreamResource

---
closes #2339
ref #3610 (an older version of this PR)

### CODE REVIEW

**Comment by ry** [comment]
@kt3k Is this working now? It looks like it. Please give an update : )

**Comment by kt3k** [comment]
@ry Yes, this is ready for review now. PTAL.

cc @bartlomieju @kevinkassimo

File: cli/js/signal_test.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+import { testPerm, assert, assertEquals, assertThrows } from "./test_util.ts";
+
+function defer(n: number): Promise<void> {
+  return new Promise((resolve, _) => {
+    setTimeout(resolve, n);
+  });
+}
+
+if (Deno.build.os === "win") {
+  testPerm({ run: true }, async function signalsNotImplemented(): Promise<
```
Comment (bartlomieju) [inline cli/js/signal_test.ts]: Why is `run` permission needed here?

File: cli/js/signals.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  private async pollSignal(): Promise<void> {
+    const { done } = await sendAsync(dispatch.OP_POLL_SIGNAL, {
+      rid: this.rid
+    });
+
+    if (done) {
+      throw new DenoError(ErrorKind.NotFound, STREAM_DISPOSED_MESSAGE);
+    }
```
Comment (bartlomieju) [inline cli/js/signals.ts]: Seems superfluous... just return `done`?

File: cli/ops/signal.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+#[derive(Deserialize)]
+struct UnbindSignalArgs {
+  rid: i32,
+}
+
+#[cfg(unix)]
+#[derive(Deserialize)]
+struct PollSignalArgs {
+  rid: i32,
+}
```
Comment (bartlomieju) [inline cli/ops/signal.rs]: How about single `SignalArgs`?

File: cli/js/signals.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  [Symbol.asyncIterator](): AsyncIterator<void> {
+    return this;
+  }
+
+  dispose(): void {
+    this.disposed = true;
+    sendSync(dispatch.OP_UNBIND_SIGNAL, { rid: this.rid });
+  }
```
Comment (bartlomieju) [inline cli/js/signals.ts]: Maybe guard against additional calls to `dispose()`

File: cli/js/dispatch.ts [hunk 5]
```diff
@@ -74,6 +74,9 @@ export let OP_HOSTNAME: number;
 export let OP_OPEN_PLUGIN: number;
 export let OP_COMPILE: number;
 export let OP_TRANSPILE: number;
+export let OP_BIND_SIGNAL: number;
+export let OP_UNBIND_SIGNAL: number;
+export let OP_POLL_SIGNAL: number;
```
Comment (ry) [inline cli/js/dispatch.ts]: Would be better to call these 

OP_SIGNAL_BIND
OP_SIGNAL_UNBIND
OP_POLL_SIGNAL

File: cli/js/lib.deno_runtime.d.ts [hunk 6]
```diff
@@ -2130,6 +2130,80 @@ declare namespace Deno {
...
+    ): Promise<T | S>;
+    next(): Promise<IteratorResult<void>>;
+    [Symbol.asyncIterator](): AsyncIterator<void>;
+    dispose(): void;
+  }
+  /**
+   * Returns the stream of the given signal number. You can use it as an async
+   * iterator.
+   *
+   *     for await (const _ of Deno.signal(Deno.Signal.SIGTERM)) {
```
Comment (ry) [inline cli/js/lib.deno_runtime.d.ts]: I'm not sure about the API Deno.Signal.SIGTERM... Would `Deno.signal("SIGTERM")` be better?

Comment (kevinkassimo) [reply cli/js/lib.deno_runtime.d.ts]: @ry IMHO using enum is probably fine. This `Deno.Signal.*` is also shared with `kill()`

Comment (ry) [reply cli/js/lib.deno_runtime.d.ts]: ok - agreed

File: cli/js/signal_test.ts (hunk 1 above)
Comment (kt3k) [reply cli/js/signal_test.ts]: Sorry. A mistake. Will remove.

File: cli/js/signals.ts (hunk 4 above)
Comment (kt3k) [reply cli/js/signals.ts]: Sounds a good idea.

**Comment by axetroy** [comment]
Is now able to listen to multiple signals at once?

If it can, it must be excellent

**Comment by kt3k** [comment]
@axetroy Sorry. I haven't included that feature in this PR because this one is already a little big and I'm not so sure about the interface of that feature (`Deno.signal(Signal.SIGTERM, Signal.SIGINT)` ? or some other function in deno_std?). I'd like to duscuss about the interface of that feature in another issue first.

**Pull request merged**